server/commands.py

```diff
@@ -617,3 +631,55 @@ def ooc_cmd_mutepm(client, arg):
         client.send_host_message('You stopped receiving PMs')
     else:
         client.send_host_message('You are now receiving PMs')
+
+def ooc_cmd_ban(client, arg):
+    if not client.is_mod:
+        raise ClientError('You must be authorized to do that.')
+    ip = arg.strip()
+    if len(ip) < 7:
+        raise ArgumentError('You must specify a valid IP.')
+    try:
+        client.server.ban_manager.add_ban(ip)
+    except ServerError:
+        raise
+    hdid = client.server.client_manager.get_hdid_by_ip(ip)
+    if hdid == None:
+        client.send_host_message('Unable to locate client HDID for logging.')
+    targets = client.server.client_manager.get_targets_by_ip(ip)
+    if targets:
+        for c in targets:
+            c.disconnect()
+        client.send_host_message('Kicked {} existing client(s).'.format(len(targets)))
+    if hdid != None:
+        client.send_host_message('Added {} to the banlist.'.format(ip))
+        logger.log_server('Banned {}, {} at IP.'.format(ip, hdid), client)
+    else:
+        client.send_host_message('Added {} to the banlist.'.format(ip))
+        logger.log_server('Banned {} at IP. Could not locate client HDID.'.format(ip, hdid), client)
+
+def ooc_cmd_banhdid(client, arg):
+    if not client.is_mod:
+        raise ClientError('You must be authorized to do that.')
+    hdid = arg.strip()
+    if len(hdid) < 8:
+        raise ArgumentError('You must specify a valid HDID.')
+    try:
+        client.server.ban_manager.add_hdidban(hdid)
+    except ServerError:
+        raise
+    ip = client.server.client_manager.get_ip_by_hdid(hdid)
+    if ip == None:
+        client.send_host_message('Unable to locate client IP for logging.')
+    else:
+        client.server.ban_manager.add_ban(ip)
+    targets = client.server.client_manager.get_targets_by_hdid(hdid)
+    if targets:
+        for c in targets:
+            c.disconnect()
```

Comment:
I don't really know an elegant way of doing that, sadly. What with a disconnect already creating a popup client-side.